File: skills/adkit/scripts/ads_skill/lib/schema_test.py
from __future__ import annotations

import pytest
from pydantic import ValidationError

from ads_skill.lib.schema import Brief


def _ad_group(name: str = "Waitlist Core", root: str = "vontevo") -> dict:
    return {
        "name": name,
        "defaultBidMicros": 1_500_000,
        "responsiveSearchAd": {
            "headlines": [{"text": f"Vontevo headline {i}"} for i in range(15)],
            "descriptions": [{"text": f"Vontevo description {i}"} for i in range(4)],
            "finalUrl": "https://www.example.com/waitlist",
        },
        "keywords": [
            {"text": root, "matchType": "PHRASE"},
            {"text": root, "matchType": "EXACT"},
        ],
    }


def _valid_brief_dict() -> dict:
    return {
        "name": "vontevo-waitlist-q3",
        "version": 1,
        "campaign": {
            "name": "Vontevo Waitlist Q3",
            "budgetMicros": 10_000_000,
            "networkSettings": "search-only",
        },
        "adGroups": [_ad_group()],
    }


def test_valid_brief_parses() -> None:
    Brief.model_validate(_valid_brief_dict())


def test_brief_at_max_ad_groups_parses() -> None:
    from ads_skill.lib.schema import MAX_AD_GROUPS
    raw = _valid_brief_dict()
    raw["adGroups"] = [
        _ad_group(name=f"Ag-{i}", root=f"root-{i}") for i in range(MAX_AD_GROUPS)
    ]
    Brief.model_validate(raw)


def test_brief_over_max_ad_groups_rejected() -> None:
    from ads_skill.lib.schema import MAX_AD_GROUPS
    raw = _valid_brief_dict()
    raw["adGroups"] = [_ad_group(name=f"Ag-{i}", root=f"root-{i}") for i in range(MAX_AD_GROUPS + 1)]
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_brief_with_zero_ad_groups_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"] = []
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_duplicate_ad_group_names_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"] = [_ad_group(name="Same"), _ad_group(name="Same", root="x")]
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_unknown_top_level_key_rejected() -> None:
    raw = _valid_brief_dict()
    raw["extraField"] = "nope"
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_invalid_name_pattern_rejected() -> None:
    raw = _valid_brief_dict()
    raw["name"] = "Has_Underscores_Caps"
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_non_https_final_url_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["finalUrl"] = "http://www.example.com/waitlist"
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_display_paths_omitted_by_default() -> None:
    brief = Brief.model_validate(_valid_brief_dict())
    rsa = brief.adGroups[0].responsiveSearchAd
    assert rsa.path1 is None and rsa.path2 is None


def test_display_paths_parse_and_lowercase() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["path1"] = "Review-Replies"
    raw["adGroups"][0]["responsiveSearchAd"]["path2"] = "Free-Trial"
    brief = Brief.model_validate(raw)
    rsa = brief.adGroups[0].responsiveSearchAd
    # mixed-case input is coerced to lower case
    assert rsa.path1 == "review-replies" and rsa.path2 == "free-trial"


def test_display_path_over_15_chars_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["path1"] = "WayTooLongPathSegment"
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_display_path_with_slash_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["path1"] = "a/b"
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_display_path_with_space_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["path1"] = "free trial"
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_display_path2_without_path1_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["path2"] = "Free-Trial"
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_display_path_todo_placeholder_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["path1"] = "TODO-keyword"
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_too_few_headlines_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["headlines"] = (
        raw["adGroups"][0]["responsiveSearchAd"]["headlines"][:14]
    )
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_duplicate_headlines_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["headlines"][1] = {"text": "Vontevo headline 0"}
    with pytest.raises(ValidationError, match="headlines must be unique"):
        Brief.model_validate(raw)


def test_pinned_headline_rejected() -> None:
    # Pinning is disabled skill-wide; pin is locked to "NONE".
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["headlines"][0] = {
        "text": "Vontevo headline 0", "pin": "HEADLINE_1"
    }
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_pinned_description_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["descriptions"][0] = {
        "text": "A description ending in act now.", "pin": "DESCRIPTION_1"
    }
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_too_many_descriptions_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["responsiveSearchAd"]["descriptions"] = [
        {"text": f"D{i}"} for i in range(5)
    ]
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_version_must_be_positive() -> None:
    raw = _valid_brief_dict()
    raw["version"] = 0
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)




def test_default_bid_strategy_is_maximize_clicks() -> None:
    # new campaigns launch on Maximize Clicks to seed conversion data (cold-start escape)
    brief = Brief.model_validate(_valid_brief_dict())
    assert brief.campaign.bidStrategy == "maximize-clicks"


def test_cpc_ceiling_requires_maximize_clicks() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["bidStrategy"] = "maximize-conversions"
    raw["campaign"]["cpcBidCeilingMicros"] = 2_000_000
    with pytest.raises(ValidationError, match="cpcBidCeilingMicros only valid"):
        Brief.model_validate(raw)


def test_cpc_ceiling_ok_with_maximize_clicks() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["bidStrategy"] = "maximize-clicks"
    raw["campaign"]["cpcBidCeilingMicros"] = 2_000_000
    Brief.model_validate(raw)


def test_manual_cpc_bid_strategy_parses() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["bidStrategy"] = "manual-cpc"
    brief = Brief.model_validate(raw)
    assert brief.campaign.bidStrategy == "manual-cpc"


def test_target_cpa_requires_target_cpa_micros() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["bidStrategy"] = "target-cpa"
    with pytest.raises(ValidationError, match="targetCpaMicros"):
        Brief.model_validate(raw)


def test_target_cpa_with_target_cpa_micros_ok() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["bidStrategy"] = "target-cpa"
    raw["campaign"]["targetCpaMicros"] = 5_000_000
    Brief.model_validate(raw)


def test_target_roas_requires_target_roas() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["bidStrategy"] = "target-roas"
    with pytest.raises(ValidationError, match="targetRoas"):
        Brief.model_validate(raw)


def test_target_roas_with_target_roas_ok() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["bidStrategy"] = "target-roas"
    raw["campaign"]["targetRoas"] = 4.0
    Brief.model_validate(raw)


def test_target_cpa_micros_rejected_when_strategy_does_not_need_it() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["bidStrategy"] = "maximize-conversions"
    raw["campaign"]["targetCpaMicros"] = 5_000_000
    with pytest.raises(ValidationError, match="targetCpaMicros only valid"):
        Brief.model_validate(raw)


def test_unknown_bid_strategy_rejected() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["bidStrategy"] = "secret-sauce"
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def _sitelink(text: str = "How It Works") -> dict:
    return {"text": text, "finalUrl": "https://www.example.com/page.html"}


def test_six_sitelinks_ok() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["sitelinks"] = [_sitelink(f"Link {i}") for i in range(6)]
    Brief.model_validate(raw)


def test_zero_sitelinks_ok_legacy() -> None:
    Brief.model_validate(_valid_brief_dict())  # no sitelinks key


def test_five_sitelinks_rejected() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["sitelinks"] = [_sitelink(f"Link {i}") for i in range(5)]
    with pytest.raises(ValidationError, match="exactly 6"):
        Brief.model_validate(raw)


def test_seven_sitelinks_rejected() -> None:
    raw = _valid_brief_dict()
    raw["campaign"]["sitelinks"] = [_sitelink(f"Link {i}") for i in range(7)]
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_sitelink_text_over_25_chars_rejected() -> None:
    raw = _valid_brief_dict()
    sl = [_sitelink(f"Link {i}") for i in range(6)]
    sl[0]["text"] = "x" * 26
    raw["campaign"]["sitelinks"] = sl
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_sitelink_one_description_rejected() -> None:
    raw = _valid_brief_dict()
    sl = [_sitelink(f"Link {i}") for i in range(6)]
    sl[0]["description1"] = "only one line"
    raw["campaign"]["sitelinks"] = sl
    with pytest.raises(ValidationError, match="both description1 and description2"):
        Brief.model_validate(raw)


def test_default_bid_over_15_dollars_rejected() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["defaultBidMicros"] = 15_000_001
    with pytest.raises(ValidationError):
        Brief.model_validate(raw)


def test_default_bid_at_15_dollars_ok() -> None:
    raw = _valid_brief_dict()
    raw["adGroups"][0]["defaultBidMicros"] = 15_000_000
    Brief.model_validate(raw)

File: skills/adkit/scripts/src/lib/schema.test.ts
import { describe, expect, it } from "vitest";
import {
  AD_GROUP_MAX_KEYWORDS,
  MAX_AD_GROUPS,
  RSAS_PER_AD_GROUP,
  displayPathPairErrors,
  parseBrief,
} from "./schema.js";

/** One RSA fixture; `variant` offsets the placeholder text so two RSAs in the same ad group don't collide. */
const rsa = (variant = 0) => ({
  headlines: Array.from({ length: 15 }, (_, i) => ({ text: `Vontevo headline ${variant}-${i}` })),
  descriptions: Array.from({ length: 4 }, (_, i) => ({ text: `Vontevo description ${variant}-${i}` })),
  finalUrl: "https://www.example.com/waitlist",
});

const adGroup = (name = "Waitlist Core", root = "vontevo") => ({
  name,
  defaultBidMicros: 1_500_000,
  responsiveSearchAds: [rsa(0), rsa(1)],
  keywords: [
    { text: root, matchType: "PHRASE" },
    { text: root, matchType: "EXACT" },
  ],
});

const validBrief = () => ({
  name: "vontevo-waitlist-q3",
  version: 1,
  campaign: {
    name: "Vontevo Waitlist Q3",
    budgetMicros: 10_000_000,
    networkSettings: "search-only",
  },
  adGroups: [adGroup()],
});

const sitelink = (text = "How It Works") => ({ text, finalUrl: "https://www.example.com/page.html" });

describe("Brief validation", () => {
  it("parses a valid brief", () => {
    expect(() => parseBrief(validBrief())).not.toThrow();
  });

  it("parses at the max ad group count", () => {
    const raw = { ...validBrief(), adGroups: Array.from({ length: MAX_AD_GROUPS }, (_, i) => adGroup(`Ag-${i}`, `root-${i}`)) };
    expect(() => parseBrief(raw)).not.toThrow();
  });

  it("rejects over the max ad group count", () => {
    const raw = { ...validBrief(), adGroups: Array.from({ length: MAX_AD_GROUPS + 1 }, (_, i) => adGroup(`Ag-${i}`, `root-${i}`)) };
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects zero ad groups", () => {
    expect(() => parseBrief({ ...validBrief(), adGroups: [] })).toThrow();
  });

  it("parses at the per-ad-group keyword cap and rejects one over", () => {
    const kw = (n: number) => Array.from({ length: n }, (_, i) => ({ text: `kw-${i}`, matchType: "PHRASE" }));
    const withKw = (n: number) => ({ ...validBrief(), adGroups: [{ ...adGroup(), keywords: kw(n) }] });
    expect(() => parseBrief(withKw(AD_GROUP_MAX_KEYWORDS))).not.toThrow();
    expect(() => parseBrief(withKw(AD_GROUP_MAX_KEYWORDS + 1))).toThrow();
  });

  it("rejects duplicate ad group names", () => {
    const raw = { ...validBrief(), adGroups: [adGroup("Same"), adGroup("Same", "x")] };
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects an unknown top-level key", () => {
    expect(() => parseBrief({ ...validBrief(), extraField: "nope" })).toThrow();
  });

  it("rejects an invalid name pattern", () => {
    expect(() => parseBrief({ ...validBrief(), name: "Has_Underscores_Caps" })).toThrow();
  });

  it("rejects a non-https final URL", () => {
    const raw = validBrief();
    raw.adGroups[0].responsiveSearchAds[0].finalUrl = "http://www.example.com/waitlist";
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects version < 1", () => {
    expect(() => parseBrief({ ...validBrief(), version: 0 })).toThrow();
  });
});

describe("RSA cardinality (exactly RSAS_PER_AD_GROUP per ad group)", () => {
  it("rejects a single RSA (the pre-change shape)", () => {
    const raw = validBrief();
    raw.adGroups[0].responsiveSearchAds = [rsa(0)];
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects three RSAs", () => {
    const raw = validBrief();
    raw.adGroups[0].responsiveSearchAds = [rsa(0), rsa(1), rsa(2)];
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects the legacy singular `responsiveSearchAd` key (no silent migration)", () => {
    const raw = validBrief() as Record<string, unknown>;
    const adGroups = raw["adGroups"] as Array<Record<string, unknown>>;
    const { responsiveSearchAds, ...rest } = adGroups[0]!;
    adGroups[0] = { ...rest, responsiveSearchAd: (responsiveSearchAds as unknown[])[0] };
    expect(() => parseBrief(raw)).toThrow();
  });

  it("accepts exactly RSAS_PER_AD_GROUP RSAs", () => {
    const raw = validBrief();
    expect(raw.adGroups[0].responsiveSearchAds).toHaveLength(RSAS_PER_AD_GROUP);
    expect(() => parseBrief(raw)).not.toThrow();
  });
});

describe("display paths", () => {
  it("omits paths by default", () => {
    const brief = parseBrief(validBrief());
    const r = brief.adGroups[0].responsiveSearchAds[0];
    expect(r.path1).toBeUndefined();
    expect(r.path2).toBeUndefined();
  });

  it("parses and lowercases", () => {
    const raw = validBrief();
    Object.assign(raw.adGroups[0].responsiveSearchAds[0], { path1: "Review-Replies", path2: "Free-Trial" });
    const brief = parseBrief(raw);
    const r = brief.adGroups[0].responsiveSearchAds[0];
    expect(r.path1).toBe("review-replies");
    expect(r.path2).toBe("free-trial");
  });

  it("rejects a path over 15 chars", () => {
    const raw = validBrief();
    Object.assign(raw.adGroups[0].responsiveSearchAds[0], { path1: "WayTooLongPathSegment" });
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects a path with a slash", () => {
    const raw = validBrief();
    Object.assign(raw.adGroups[0].responsiveSearchAds[0], { path1: "a/b" });
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects a path with a space", () => {
    const raw = validBrief();
    Object.assign(raw.adGroups[0].responsiveSearchAds[0], { path1: "free trial" });
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects path2 without path1", () => {
    const raw = validBrief();
    Object.assign(raw.adGroups[0].responsiveSearchAds[0], { path2: "Free-Trial" });
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects a TODO placeholder path", () => {
    const raw = validBrief();
    Object.assign(raw.adGroups[0].responsiveSearchAds[0], { path1: "TODO-keyword" });
    expect(() => parseBrief(raw)).toThrow();
  });
});

describe("displayPathPairErrors (shared create/update rules)", () => {
  it("accepts an absent pair", () => {
    expect(displayPathPairErrors(undefined, undefined)).toEqual([]);
  });

  it("accepts a valid pair", () => {
    expect(displayPathPairErrors("demo", "trial")).toEqual([]);
  });

  it("rejects a non-string segment", () => {
    expect(displayPathPairErrors(42, undefined).some((m) => m.includes("path1 must be a string"))).toBe(true);
  });

  it("prefixes each message", () => {
    const errs = displayPathPairErrors("free trial", undefined, "ad 9: ");
    expect(errs.every((m) => m.startsWith("ad 9: "))).toBe(true);
  });

  it("rejects path2 without path1", () => {
    expect(displayPathPairErrors(undefined, "trial").some((m) => m.includes("path2 requires path1"))).toBe(true);
  });
});

describe("RSA assets", () => {
  it("rejects too few headlines", () => {
    const raw = validBrief();
    raw.adGroups[0].responsiveSearchAds[0].headlines = raw.adGroups[0].responsiveSearchAds[0].headlines.slice(0, 14);
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects duplicate headlines", () => {
    const raw = validBrief();
    raw.adGroups[0].responsiveSearchAds[0].headlines[1] = { text: raw.adGroups[0].responsiveSearchAds[0].headlines[0]!.text };
    expect(() => parseBrief(raw)).toThrow(/headlines must be unique/);
  });

  it("rejects a pinned headline", () => {
    const raw = validBrief();
    raw.adGroups[0].responsiveSearchAds[0].headlines[0] = { text: "Vontevo headline 0", pin: "HEADLINE_1" } as never;
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects a pinned description", () => {
    const raw = validBrief();
    raw.adGroups[0].responsiveSearchAds[0].descriptions[0] = { text: "A description ending in act now.", pin: "DESCRIPTION_1" } as never;
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects too many descriptions", () => {
    const raw = validBrief();
    raw.adGroups[0].responsiveSearchAds[0].descriptions = Array.from({ length: 5 }, (_, i) => ({ text: `D${i}` }));
    expect(() => parseBrief(raw)).toThrow();
  });

  it("validates each RSA in the pair independently (second RSA's rule violation still rejects)", () => {
    const raw = validBrief();
    raw.adGroups[0].responsiveSearchAds[1].headlines = raw.adGroups[0].responsiveSearchAds[1].headlines.slice(0, 14);
    expect(() => parseBrief(raw)).toThrow();
  });
});

describe("bid strategy", () => {
  it("defaults to maximize-clicks", () => {
    expect(parseBrief(validBrief()).campaign.bidStrategy).toBe("maximize-clicks");
  });

  it("rejects a cpc ceiling without maximize-clicks", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { bidStrategy: "maximize-conversions", cpcBidCeilingMicros: 2_000_000 });
    expect(() => parseBrief(raw)).toThrow(/cpcBidCeilingMicros only valid/);
  });

  it("accepts a cpc ceiling with maximize-clicks", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { bidStrategy: "maximize-clicks", cpcBidCeilingMicros: 2_000_000 });
    expect(() => parseBrief(raw)).not.toThrow();
  });

  it("parses manual-cpc", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { bidStrategy: "manual-cpc" });
    expect(parseBrief(raw).campaign.bidStrategy).toBe("manual-cpc");
  });

  it("requires targetCpaMicros for target-cpa", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { bidStrategy: "target-cpa" });
    expect(() => parseBrief(raw)).toThrow(/targetCpaMicros/);
  });

  it("accepts target-cpa with targetCpaMicros", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { bidStrategy: "target-cpa", targetCpaMicros: 5_000_000 });
    expect(() => parseBrief(raw)).not.toThrow();
  });

  it("requires targetRoas for target-roas", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { bidStrategy: "target-roas" });
    expect(() => parseBrief(raw)).toThrow(/targetRoas/);
  });

  it("accepts target-roas with targetRoas", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { bidStrategy: "target-roas", targetRoas: 4.0 });
    expect(() => parseBrief(raw)).not.toThrow();
  });

  it("rejects targetCpaMicros when the strategy does not need it", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { bidStrategy: "maximize-conversions", targetCpaMicros: 5_000_000 });
    expect(() => parseBrief(raw)).toThrow(/targetCpaMicros only valid/);
  });

  it("rejects an unknown bid strategy", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { bidStrategy: "secret-sauce" });
    expect(() => parseBrief(raw)).toThrow();
  });
});

describe("sitelinks & callouts", () => {
  it("accepts six sitelinks", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { sitelinks: Array.from({ length: 6 }, (_, i) => sitelink(`Link ${i}`)) });
    expect(() => parseBrief(raw)).not.toThrow();
  });

  it("accepts zero sitelinks (legacy)", () => {
    expect(() => parseBrief(validBrief())).not.toThrow();
  });

  it("rejects five sitelinks", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { sitelinks: Array.from({ length: 5 }, (_, i) => sitelink(`Link ${i}`)) });
    expect(() => parseBrief(raw)).toThrow(/exactly 6/);
  });

  it("rejects seven sitelinks", () => {
    const raw = validBrief();
    Object.assign(raw.campaign, { sitelinks: Array.from({ length: 7 }, (_, i) => sitelink(`Link ${i}`)) });
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects sitelink text over 25 chars", () => {
    const raw = validBrief();
    const sl = Array.from({ length: 6 }, (_, i) => sitelink(`Link ${i}`));
    sl[0].text = "x".repeat(26);
    Object.assign(raw.campaign, { sitelinks: sl });
    expect(() => parseBrief(raw)).toThrow();
  });

  it("rejects a sitelink with one description", () => {
    const raw = validBrief();
    const sl = Array.from({ length: 6 }, (_, i) => sitelink(`Link ${i}`));
    Object.assign(sl[0], { description1: "only one line" });
    Object.assign(raw.campaign, { sitelinks: sl });
    expect(() => parseBrief(raw)).toThrow(/both description1 and description2/);
  });
});

describe("default bid ceiling", () => {
  it("rejects a default bid over $15", () => {
    const raw = validBrief();
    raw.adGroups[0].defaultBidMicros = 15_000_001;
    expect(() => parseBrief(raw)).toThrow();
  });

  it("accepts a default bid at $15", () => {
    const raw = validBrief();
    raw.adGroups[0].defaultBidMicros = 15_000_000;
    expect(() => parseBrief(raw)).not.toThrow();
  });
});
